kani-driver/src/exe_trace.rs

```diff
@@ -0,0 +1,288 @@
+}
+
+/// Read the CBMC output, parse it as a JSON object, and extract the deterministic values.
+mod parser {
+    use crate::util::append_path;
+    use serde_json::Value;
+    use std::fs::File;
+    use std::io::BufReader;
+    use std::path::Path;
+
+    /// Extract deterministic values from a failing harness.
+    pub fn get_det_vals(file: &Path) -> (Vec<Vec<u8>>, Vec<String>) {
+        let output_filename = append_path(file, "cbmc_output");
+        let cbmc_out = get_cbmc_out(&output_filename);
+        handle_cbmc_out(&cbmc_out)
+    }
+
+    /// Read in the CBMC results file and deserialize it to a JSON object.
+    fn get_cbmc_out(results_filename: &Path) -> Value {
+        let results_file = File::open(results_filename).unwrap();
+        let reader = BufReader::new(results_file);
+        serde_json::from_reader(reader).unwrap()
+    }
+
+    /// The first-level JSON object parser. This extracts the result message.
+    fn handle_cbmc_out(cbmc_out: &Value) -> (Vec<Vec<u8>>, Vec<String>) {
+        let mut det_vals: Vec<Vec<u8>> = Vec::new();
+        let mut interp_det_vals: Vec<String> = Vec::new();
+        for general_msg in cbmc_out.as_array().unwrap() {
+            let result_msg = &general_msg["result"];
+            if !result_msg.is_null() {
+                for result_val in result_msg.as_array().unwrap() {
+                    handle_result(result_val, &mut det_vals, &mut interp_det_vals);
+                }
+            }
+        }
+        (det_vals, interp_det_vals)
+    }
+
+    /// The second-level JSON object parser. This extracts the traces of failing assertions.
+    fn handle_result(
+        result_val: &Value,
+        det_vals: &mut Vec<Vec<u8>>,
+        interp_det_vals: &mut Vec<String>,
+    ) {
+        let desc = result_val["description"].to_string();
+        let status = result_val["status"].to_string();
+
+        if desc.contains("assertion failed") && status == "\"FAILURE\"" {
+            for trace_pt in result_val["trace"].as_array().unwrap() {
+                handle_trace_pt(trace_pt, det_vals, interp_det_vals);
+            }
+        }
+    }
+
+    /// The third-level of JSON object parser. This extracts individual bytes from kani::any_raw calls.
+    fn handle_trace_pt(
+        trace_pt: &Value,
+        det_vals: &mut Vec<Vec<u8>>,
+        interp_det_vals: &mut Vec<String>,
+    ) {
+        if let (
+            Some(step_type),
+            Some(lhs),
+            Some(func),
+            Some(bit_det_val),
+            Some(interp_det_val),
+            Some(width_u64),
+        ) = (
+            trace_pt["stepType"].as_str(),
+            trace_pt["lhs"].as_str(),
+            trace_pt["sourceLocation"]["function"].as_str(),
+            trace_pt["value"]["binary"].as_str(),
+            trace_pt["value"]["data"].as_str(),
+            trace_pt["value"]["width"].as_u64(),
+        ) && step_type == "assignment" && lhs == "var_0" && func.starts_with("kani::any_raw_internal") {
```

Comment:
Can you at least break this down into to `if` statements? That is quite hard to read.

```
if step_type == "assignment" && lhs == "var_0" && func.starts_with("kani::any_raw_internal") {
  if let ... {
  }
}

Reply:
Done